Libraries/Opc.Ua.Server/Diagnostics/CustomNodeManager.cs

```diff
@@ -4621,65 +4629,41 @@ private class CacheEntry
         /// </summary>
         protected NodeState LookupNodeInComponentCache(ISystemContext context, NodeHandle handle)
         {
-            lock (Lock)
-            {
-                if (m_componentCache == null)
-                {
-                    return null;
-                }
+            bool componentPathExists = !string.IsNullOrEmpty(handle.ComponentPath);
+            NodeId nodeId = componentPathExists ? handle.RootId : handle.NodeId;
 
-                CacheEntry entry = null;
-
-                if (!String.IsNullOrEmpty(handle.ComponentPath))
+            if (m_componentCache?.TryGetValue(nodeId, out CacheEntry entry) == true)
+            {
+                if (componentPathExists)
                 {
-                    if (m_componentCache.TryGetValue(handle.RootId, out entry))
-                    {
-                        return entry.Entry.FindChildBySymbolicName(context, handle.ComponentPath);
-                    }
+                    return entry.Entry.FindChildBySymbolicName(context, handle.ComponentPath);
                 }
                 else
                 {
-                    if (m_componentCache.TryGetValue(handle.NodeId, out entry))
-                    {
-                        return entry.Entry;
-                    }
+                    return entry.Entry;
                 }
-
-                return null;
             }
+
+            return null;
         }
 
         /// <summary>
         /// Removes a reference to a component in thecache.
         /// </summary>
         protected void RemoveNodeFromComponentCache(ISystemContext context, NodeHandle handle)
         {
-            lock (Lock)
-            {
-                if (handle == null)
-                {
-                    return;
-                }
+            bool componentPathExists = !string.IsNullOrEmpty(handle.ComponentPath);
+            NodeId nodeId = componentPathExists ? handle.RootId : handle.NodeId;
 
-                if (m_componentCache != null)
+            if (m_componentCache?.TryGetValue(nodeId, out CacheEntry entry) == true)
+            {
+                lock (Lock)
                 {
-                    NodeId nodeId = handle.NodeId;
+                    int refCount = Interlocked.Decrement(ref entry.RefCount);
 
-                    if (!String.IsNullOrEmpty(handle.ComponentPath))
+                    if (refCount == 0)
```

Comment:
how to make shure the refCount is still 0 when the add methods are not locked any more

Reply:
solved with: https://devblogs.microsoft.com/pfxteam/little-known-gems-atomic-conditional-removals-from-concurrentdictionary/